src/unix/pty.cc

```diff
@@ -242,85 +233,108 @@ NAN_METHOD(PtyFork) {
+
+  pid_t pid;
+  auto error = posix_spawn(&pid, helper_path, &acts, &attrs, argv, env);
+
+  posix_spawn_file_actions_destroy(&acts);
+  posix_spawnattr_destroy(&attrs);
+  close(comms_pipe[1]);
+
   // reenable signals
   pthread_sigmask(SIG_SETMASK, &oldmask, NULL);
 
-  if (pid) {
-    for (i = 0; i < argl; i++) free(argv[i]);
-    delete[] argv;
-    for (i = 0; i < envc; i++) free(env[i]);
-    delete[] env;
-    free(cwd);
+  if (error) {
+    perror("posix_spawn failed");
+    return Nan::ThrowError("posix_spawn(3) failed.");
   }
 
-  switch (pid) {
-    case -1:
-      return Nan::ThrowError("forkpty(3) failed.");
-    case 0:
-      if (strlen(cwd)) {
-        if (chdir(cwd) == -1) {
-          perror("chdir(2) failed.");
-          _exit(1);
-        }
-      }
-
-      if (uid != -1 && gid != -1) {
-        if (setgid(gid) == -1) {
-          perror("setgid(2) failed.");
-          _exit(1);
-        }
-        if (setuid(uid) == -1) {
-          perror("setuid(2) failed.");
-          _exit(1);
-        }
-      }
-
-      pty_execvpe(argv[0], argv, env);
-
-      perror("execvp(3) failed.");
-      _exit(1);
-    default:
-      if (pty_nonblock(master) == -1) {
-        return Nan::ThrowError("Could not set master fd to nonblocking.");
-      }
-
-      v8::Local<v8::Object> obj = Nan::New<v8::Object>();
-      Nan::Set(obj,
-        Nan::New<v8::String>("fd").ToLocalChecked(),
-        Nan::New<v8::Number>(master));
-      Nan::Set(obj,
-        Nan::New<v8::String>("pid").ToLocalChecked(),
-        Nan::New<v8::Number>(pid));
-      Nan::Set(obj,
-        Nan::New<v8::String>("pty").ToLocalChecked(),
-        Nan::New<v8::String>(ptsname(master)).ToLocalChecked());
-
-      pty_baton *baton = new pty_baton();
-      baton->exit_code = 0;
-      baton->signal_code = 0;
-      baton->cb.Reset(v8::Local<v8::Function>::Cast(info[9]));
-      baton->pid = pid;
-      baton->async.data = baton;
-
-      uv_async_init(uv_default_loop(), &baton->async, pty_after_waitpid);
-
-      uv_thread_create(&baton->tid, pty_waitpid, static_cast<void*>(baton));
-
-      return info.GetReturnValue().Set(obj);
+  comm_send_int(comms_pipe[0], COMM_MSG_PATH);
+  comm_send_str(comms_pipe[0], argv[0]);
+  comm_send_int(comms_pipe[0], COMM_MSG_ARGV);
+  comm_send_str_array(comms_pipe[0], argv);
+  comm_send_int(comms_pipe[0], COMM_MSG_ENV);
+  comm_send_str_array(comms_pipe[0], env);
+  if (strlen(cwd)) {
+    comm_send_int(comms_pipe[0], COMM_MSG_CWD);
+    comm_send_str(comms_pipe[0], cwd);
+  }
+  if (uid != -1) {
+    comm_send_int(comms_pipe[0], COMM_MSG_UID);
+    comm_send_int(comms_pipe[0], uid);
   }
+  if (gid != -1) {
+    comm_send_int(comms_pipe[0], COMM_MSG_GID);
+    comm_send_int(comms_pipe[0], gid);
+  }
+  comm_send_int(comms_pipe[0], COMM_MSG_GO_FOR_LAUNCH);
 
-  return info.GetReturnValue().SetUndefined();
+  int result;
+  auto bytes_read = recv(comms_pipe[0], &result, sizeof(result), 0);
```

Comment:
With a pipe instead of a socket this is abit different:
- either just use a blocking read on the pipe, when returning 0 we are done and can delete the fd
- since blocking is prolly unwanted here - move the fd into the waitpid thread, and do the blocking read/delete there

I think the onExit event needs some revamping with `posix_spawn`, since we dont get exec errors synchronously anymore (they will end up in the waitpid path as well). Maybe we need an async onExecError event now?